storage/src/main/java/org/apache/kafka/server/log/internals/ProducerStateEntry.java

```diff
@@ -0,0 +1,143 @@
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.apache.kafka.server.log.internals;
+
+import org.apache.kafka.common.record.RecordBatch;
+
+import java.util.ArrayList;
+import java.util.Collections;
+import java.util.List;
+import java.util.Optional;
+import java.util.OptionalLong;
+import java.util.stream.Stream;
+
+/**
+ * The batchMetadata is ordered such that the batch with the lowest sequence is at the head of the queue while the
+ * batch with the highest sequence is at the tail of the queue. We will retain at most {@link ProducerStateEntry#NUM_BATCHES_TO_RETAIN}
+ * elements in the queue. When the queue is at capacity, we remove the first element to make space for the incoming batch.
+ */
+public class ProducerStateEntry {
+    public static final int NUM_BATCHES_TO_RETAIN = 5;
+    public final long producerId;
+    private final List<BatchMetadata> batchMetadata;
+    private short producerEpoch;
+    public int coordinatorEpoch;
+    public long lastTimestamp;
+    public OptionalLong currentTxnFirstOffset;
+
+    public ProducerStateEntry(long producerId) {
+        this(producerId, new ArrayList<>(), RecordBatch.NO_PRODUCER_EPOCH, -1, RecordBatch.NO_TIMESTAMP, OptionalLong.empty());
+    }
+
+    public ProducerStateEntry(long producerId, short producerEpoch, int coordinatorEpoch, long lastTimestamp, OptionalLong currentTxnFirstOffset) {
+        this(producerId, new ArrayList<>(), producerEpoch, coordinatorEpoch, lastTimestamp, currentTxnFirstOffset);
+    }
+
+    public ProducerStateEntry(long producerId, List<BatchMetadata> batchMetadata, short producerEpoch, int coordinatorEpoch, long lastTimestamp, OptionalLong currentTxnFirstOffset) {
+        this.producerId = producerId;
+        this.batchMetadata = batchMetadata;
+        this.producerEpoch = producerEpoch;
+        this.coordinatorEpoch = coordinatorEpoch;
+        this.lastTimestamp = lastTimestamp;
+        this.currentTxnFirstOffset = currentTxnFirstOffset;
+    }
+
+    public int firstSeq() {
+        return isEmpty() ? RecordBatch.NO_SEQUENCE : batchMetadata.get(0).firstSeq();
+    }
+
+
+    public long firstDataOffset() {
+        return isEmpty() ? -1L : batchMetadata.get(0).firstOffset();
+    }
+
+    public int lastSeq() {
+        return isEmpty() ? RecordBatch.NO_SEQUENCE : batchMetadata.get(batchMetadata.size() - 1).lastSeq;
+    }
+
+    public long lastDataOffset() {
+        return isEmpty() ? -1L : batchMetadata.get(batchMetadata.size() - 1).lastOffset;
+    }
+
+    public int lastOffsetDelta() {
+        return isEmpty() ? 0 : batchMetadata.get(batchMetadata.size() - 1).offsetDelta;
```

Comment:
You can add a private `lastOrFail` method to simplify these `batchMetadata.get(batchMetadata.size() - 1)` calls. For symmetry, you can also add a `firstOrFail` method.